### PR TITLE
Implement gRPC provision plugins

### PR DESCRIPTION
**Pull request by markgoddard**
- **Remove unnecessary types in values tests**
- **Export helm.MergeMaps**
- **Implement gRPC provision plugin stubs with data source host plugin**
- **Refactor plugins configuration**
- **Rework gRPC plugin loading**
- **Rename gRPC plugin serving arguments**
  From 'data-source serve' to 'plugin serve'.

Fixes: #56 
Depends-on: https://github.com/cofide/cofide-api-sdk/pull/14

### CODE REVIEW

File: cmd/cofidectl/cmd/down.go [hunk 1]
```diff
@@ -29,12 +29,15 @@ func (d *DownCommand) DownCmd() *cobra.Command {
...
+			ds, err := d.cmdCtx.PluginManager.GetDataSource(cmd.Context())
 			if err != nil {
 				return err
 			}
 
-			provision := d.cmdCtx.PluginManager.GetProvision()
+			provision, err := d.cmdCtx.PluginManager.GetProvision(cmd.Context())
+			if err != nil {
+				return err
+			}
```
Comment (nialdaly) [inline cmd/cofidectl/cmd/down.go:40]: Definitely a nit but maybe we could add a newline here?

File: internal/pkg/config/config.go [hunk 2]
```diff
@@ -7,42 +7,47 @@ import (
 	"buf.build/go/protoyaml"
 	attestation_policy_proto "github.com/cofide/cofide-api-sdk/gen/go/proto/attestation_policy/v1alpha1"
 	config_proto "github.com/cofide/cofide-api-sdk/gen/go/proto/config/v1alpha1"
+	pluginspb "github.com/cofide/cofide-api-sdk/gen/go/proto/plugins/v1alpha1"
```
Comment (nialdaly) [inline internal/pkg/config/config.go:10]: Nit: this alias could be `plugins_proto`, similar to how others are named perhaps?

File: internal/pkg/config/config_test.go [hunk 3]
```diff
@@ -7,11 +7,11 @@ import (
 	"testing"
 
 	attestation_policy_proto "github.com/cofide/cofide-api-sdk/gen/go/proto/attestation_policy/v1alpha1"
+	pluginspb "github.com/cofide/cofide-api-sdk/gen/go/proto/plugins/v1alpha1"
```
Comment (nialdaly) [inline internal/pkg/config/config_test.go:10]: ditto

File: internal/pkg/config/loader_test.go [hunk 4]
```diff
@@ -29,21 +27,14 @@ func TestFileLoaderNonExistentConfig(t *testing.T) {
 	loader := NewFileLoader(filepath.Join(tempDir, "non-existent.yaml"))
 
 	gotExists, err := loader.Exists()
-	if err != nil {
-		t.Errorf("FileLoader.Exists() error = %v", err)
-	} else if gotExists {
-		t.Errorf("FileLoader.Exists() returned true")
-	}
+	require.NoError(t, err, err)
```
Comment (nialdaly) [inline internal/pkg/config/loader_test.go:30]: 👌

File: internal/pkg/config/loader_test.go [hunk 5]
```diff
@@ -133,68 +106,45 @@ func TestMemoryLoaderImplementsLoader(t *testing.T) {
...
-		t.Errorf("MemoryLoader.Exists() returned true")
-	}
+	require.NoError(t, err, err)
+	assert.False(t, gotExists, "MemoryLoader.Exists() returned true")
 
 	_, gotErr := loader.Read()
-	if gotErr == nil {
-		t.Fatal("MemoryLoader.Read() did not return error")
-	}
+	require.Error(t, gotErr, gotErr)
```
Comment (nialdaly) [inline internal/pkg/config/loader_test.go:116]: Is this check correct?

File: internal/pkg/proto/proto.go [hunk 6]
```diff
@@ -12,6 +12,7 @@ import (
 	ap_binding_proto "github.com/cofide/cofide-api-sdk/gen/go/proto/ap_binding/v1alpha1"
 	attestation_policy_proto "github.com/cofide/cofide-api-sdk/gen/go/proto/attestation_policy/v1alpha1"
 	federation_proto "github.com/cofide/cofide-api-sdk/gen/go/proto/federation/v1alpha1"
+	pluginspb "github.com/cofide/cofide-api-sdk/gen/go/proto/plugins/v1alpha1"
```
Comment (nialdaly) [inline internal/pkg/proto/proto.go:15]: ditto

File: internal/pkg/proto/proto.go [hunk 7]
```diff
@@ -48,6 +49,28 @@ func CloneAPBinding(binding *ap_binding_proto.APBinding) (*ap_binding_proto.APBi
 	}
 }
 
+func CloneFederation(federation *federation_proto.Federation) (*federation_proto.Federation, error) {
+	if clone, ok := proto.Clone(federation).(*federation_proto.Federation); !ok {
+		return nil, fmt.Errorf("bug: type assertion failed for federation %s-%s", federation.From, federation.To)
+	} else {
+		if clone == federation {
+			return nil, fmt.Errorf("bug: federation %s-%s clones are the same", federation.To, federation.To)
```
Comment (nialdaly) [inline internal/pkg/proto/proto.go:57]: Should this be `fmt.Errorf("bug: federation %s-%s clones are the same", federation.From, federation.To)` perhaps?

Comment (nialdaly) [reply internal/pkg/proto/proto.go:57]:
```suggestion
			return nil, fmt.Errorf("bug: federation %s-%s clones are the same", federation.From, federation.To)
```

Comment (nialdaly) [inline internal/pkg/proto/proto.go:57]: Is this error message correct?

File: internal/pkg/test/fixtures/fixtures.go [hunk 8]
```diff
@@ -9,6 +9,7 @@ import (
 	ap_binding_proto "github.com/cofide/cofide-api-sdk/gen/go/proto/ap_binding/v1alpha1"
 	attestation_policy_proto "github.com/cofide/cofide-api-sdk/gen/go/proto/attestation_policy/v1alpha1"
 	federation_proto "github.com/cofide/cofide-api-sdk/gen/go/proto/federation/v1alpha1"
+	pluginspb "github.com/cofide/cofide-api-sdk/gen/go/proto/plugins/v1alpha1"
```
Comment (nialdaly) [inline internal/pkg/test/fixtures/fixtures.go:12]: ditto

File: pkg/plugin/local/local.go [hunk 9]
```diff
@@ -36,7 +37,7 @@ func NewLocalDataSource(loader config.Loader) (*LocalDataSource, error) {
 	return lds, nil
 }
 
-func (lds *LocalDataSource) Validate() error {
+func (lds *LocalDataSource) Validate(_ context.Context) error {
```
Comment (nialdaly) [inline pkg/plugin/local/local.go:40]: 👍

File: pkg/plugin/manager/manager.go [hunk 10]
```diff
@@ -4,161 +4,249 @@
...
 
 import (
+	"context"
 	"errors"
 	"fmt"
 	"maps"
 	"os"
 	"os/exec"
 
+	pluginspb "github.com/cofide/cofide-api-sdk/gen/go/proto/plugins/v1alpha1"
```
Comment (nialdaly) [inline pkg/plugin/manager/manager.go:14]: ditto

File: pkg/plugin/provision/plugin.go [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+
+// wrapError returns a clientError if the provided error is a gRPC status, or the original error otherwise.
+func wrapError(err error) error {
+	if err == nil {
+		return nil
+	}
+	if status, ok := status.FromError(err); ok {
+		return &clientError{status: status}
+	} else {
+		return err
```
Comment (nialdaly) [inline pkg/plugin/provision/plugin.go:154]: Maybe we could just return here like:
```
	if status, ok := status.FromError(err); ok {
		return &clientError{status: status}
	}
	
        return err
```

File: pkg/plugin/provision/plugin.go [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+
+package provision
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	cofidectl_proto "github.com/cofide/cofide-api-sdk/gen/go/proto/cofidectl_plugin/v1alpha1"
+	provisionpb "github.com/cofide/cofide-api-sdk/gen/go/proto/provision_plugin/v1alpha1"
```
Comment (nialdaly) [inline pkg/plugin/provision/plugin.go:12]: ditto

File: pkg/plugin/manager/manager.go [hunk 13]
```diff
@@ -4,161 +4,249 @@
...
-	pm.client = client
-	return ds, nil
+	return pm.provision, nil
+}
+
+// loadGrpcPlugin loads a gRPC plugin.
+// The gRPC plugin may provide one or more cofidectl plugins (e.g. data source, provision), and
+// all cofidectl plugins configured to use this gRPC plugin will be loaded in a single plugin
+// client and server process.
+func (pm *PluginManager) loadGrpcPlugin(ctx context.Context, pluginName string, pluginCfg *pluginspb.Plugins) error {
```
Comment (nialdaly) [inline pkg/plugin/manager/manager.go:195]: Nit: Could this be renamed to `loadGrpcPlugin` perhaps, following this [suggestion](https://go.dev/wiki/CodeReviewComments#initialisms)?

File: pkg/plugin/manager/manager_test.go [hunk 14]
```diff
@@ -4,12 +4,18 @@
 package manager
 
 import (
+	"context"
 	"errors"
 	"testing"
 
+	pluginspb "github.com/cofide/cofide-api-sdk/gen/go/proto/plugins/v1alpha1"
```
Comment (nialdaly) [inline pkg/plugin/manager/manager_test.go:11]: ditto

File: pkg/plugin/manager/manager_test.go [hunk 15]
```diff
@@ -27,73 +33,57 @@ func newFakeGrpcDataSource(t *testing.T, configLoader config.Loader) *fakeGrpcDa
 	return &fakeGrpcDataSource{LocalDataSource: *lds}
 }
 
+type fakeGrpcProvision struct {
```
Comment (nialdaly) [inline pkg/plugin/manager/manager_test.go:36]: ditto - perhaps `fakeGRPCProvision`?

File: pkg/plugin/manager/manager_test.go [hunk 16]
```diff
@@ -27,73 +33,57 @@ func newFakeGrpcDataSource(t *testing.T, configLoader config.Loader) *fakeGrpcDa
...
+	spirehelm.SpireHelm
+}
+
+func newFakeGrpcProvision() *fakeGrpcProvision {
```
Comment (nialdaly) [inline pkg/plugin/manager/manager_test.go:40]: ditto

File: internal/pkg/config/config.go (hunk 2 above)
Comment (markgoddard) [reply internal/pkg/config/config.go:10]: The underscores in the proto import renames [aren't idiomatic](https://go.dev/blog/package-names#package-paths).

> When renaming an imported package, the local name should follow the same guidelines as package names (lower case, no under_scores or mixedCaps).

I figured for existing packages we can keep them for now but any new ones should probably use a different scheme. I was thinking of doing a big find/replace for the older ones when things are a bit quieter.

The `pb` suffix seems to be quite commonly used, at least in Google, and is nice and short (https://pkg.go.dev/google.golang.org/protobuf/types/known/anypb, https://pkg.go.dev/google.golang.org/protobuf/types/known/structpb). Spire uses a <package><version> format, e.g. `bundlev1`.

File: internal/pkg/config/loader_test.go (hunk 5 above)
Comment (markgoddard) [reply internal/pkg/config/loader_test.go:116]: Yes - previous code failed if there was no error.

File: internal/pkg/proto/proto.go (hunk 7 above)
Comment (markgoddard) [reply internal/pkg/proto/proto.go]: Good catch. This code was actually just moved to group the Clone functions together but I'll fix it here.

File: pkg/plugin/manager/manager.go (hunk 13 above)
Comment (markgoddard) [reply pkg/plugin/manager/manager.go:195]: oops, missed this one

**Review by nialdaly — APPROVED** [review]

**Pull request merged**